performance/multi-node/rdm_one_sided.c

```diff
@@ -0,0 +1,747 @@
+	ret = fi_getinfo(CT_FIVERSION, NULL, NULL, flags, hints, &fi);
+	if (ret) {
+		ct_print_fi_error("fi_getinfo", ret);
+		return ret;
+	}
+
+	/* Open fabric */
+	ret = fi_fabric(fi->fabric_attr, &fab, NULL);
+	if (ret) {
+		ct_print_fi_error("fi_fabric", ret);
+		goto err1;
+	}
+
+
+	if (!thread_safe) {
+		fi->domain_attr->threading = FI_THREAD_COMPLETION;
+		fi->domain_attr->data_progress = FI_PROGRESS_MANUAL;
+		fi->domain_attr->control_progress = FI_PROGRESS_MANUAL;
+	}
+
+	/* Open domain */
+	ret = fi_domain(fab, fi, &dom, NULL);
+	if (ret) {
+		ct_print_fi_error("fi_domain", ret);
+		goto err2;
+	}
+
+	return 0;
+
+err2:
+	fi_close(&fab->fid);
+err1:
+	return ret;
+}
+
+static int init_endpoint(struct per_thread_data *ptd)
+{
+	int ret;
+
+	/* Open endpoint */
+	ret = fi_endpoint(dom, fi, &ptd->ep, NULL);
+	if (ret) {
+		ct_print_fi_error("fi_endpoint", ret);
+		goto err3;
+	}
+
+	/* Allocate endpoint resources */
+	ret = alloc_ep_res(ptd);
+	if (ret)
+		goto err4;
+
+	/* Bind EQs and AVs with endpoint */
+	ret = bind_ep_res(ptd);
+	if (ret)
+		goto err5;
+
+	return 0;
+
+err5:
+	free_ep_res(ptd);
+err4:
+	fi_close(&ptd->ep->fid);
+err3:
+	fi_close(&dom->fid);
+	return ret;
+}
+
+static int init_av(struct per_thread_data *ptd)
+{
+	void *addr;
+	size_t addrlen = 0;
+	int ret;
+
+	fi_getname(&ptd->ep->fid, NULL, &addrlen);
+	addr = malloc(addrlen);
+	assert(addr);
+	ret = fi_getname(&ptd->ep->fid, addr, &addrlen);
+	if (ret != 0) {
+		ct_print_fi_error("fi_getname", ret);
+		return ret;
+	}
+
+	ptd->addrs = malloc(numprocs * addrlen);
+	assert(ptd->addrs);
+
+	CT_Allgather(addr, addrlen, ptd->addrs);
+
+	ptd->fi_addrs = malloc(numprocs * sizeof(fi_addr_t));
+	assert(ptd->fi_addrs);
+
+	/* Insert address to the AV and get the fabric address back */
+	ret = fi_av_insert(ptd->av, ptd->addrs, numprocs, ptd->fi_addrs, 0, &ptd->fi_ctx_av);
+	if (ret != numprocs) {
+		ct_print_fi_error("fi_av_insert", ret);
+		return ret;
+	}
+
+	free(addr);
+
+	return 0;
+}
+
+int init_per_thread_data(struct per_thread_data *ptd)
+{
+	int align_size;
+	int ret;
+	buf_desc_t lbuf_desc;
+
+	ret = init_endpoint(ptd);
+	if (ret) {
+		fprintf(stderr, "Problem in endpoint initialization\n");
+		return ret;
+	}
+
+	ret = init_av(ptd);
+	if (ret) {
+		fprintf(stderr, "Problem in AV initialization\n");
+		return ret;
+	}
+
+	/* Data initialization */
+	align_size = getpagesize();
+	assert(align_size <= MAX_ALIGNMENT);
+
+	ptd->s_buf = (char *) (((unsigned long) ptd->s_buf_original + (align_size - 1)) /
+				align_size * align_size);
+	ptd->r_buf = (char *) (((unsigned long) ptd->r_buf_original + (align_size - 1)) /
+				align_size * align_size);
+
+	ret = fi_mr_reg(dom, ptd->r_buf, MYBUFSIZE, FI_REMOTE_WRITE, 0, 0, 0, &ptd->r_mr, NULL);
+	if (ret) {
+		ct_print_fi_error("fi_mr_reg", ret);
+		return -1;
+	}
+
+	lbuf_desc.addr = (uint64_t) ptd->r_buf;
+	lbuf_desc.key = fi_mr_key(ptd->r_mr);
+
+	ptd->rbuf_descs = (buf_desc_t *) malloc(numprocs * sizeof(buf_desc_t));
```

Comment:
Maybe this isn't important since these are benchmarks, but, should this be free'd?